op-node/rollup/interop/managed/system.go

```diff
@@ -378,60 +378,131 @@ func (m *ManagedMode) Reset(ctx context.Context, lUnsafe, xUnsafe, lSafe, xSafe,
 	return nil
 }
 
-// findLatestValidLocalUnsafe scans the L2 unsafe chain for the latest block with a valid L1 origin,
-// starting from its latest unsafe and going back until the target provided by the supervisor.
-// If the supervisor's target is ahead, the latest unsafe is returned.
+// findLatestValidLocalUnsafe searches and returns the latest valid block of the L2 chain
+// starting from `l2UnsafeTarget` and checking until the latest unsafe block.
 func (m *ManagedMode) findLatestValidLocalUnsafe(ctx context.Context, l2UnsafeTarget eth.BlockID) (eth.L2BlockRef, error) {
 	latestUnsafe, err := m.l2.L2BlockRefByLabel(ctx, eth.Unsafe)
 	if err != nil {
 		return eth.L2BlockRef{}, err
 	}
+
 	logger := m.log.New("target", l2UnsafeTarget, "latestUnsafe", latestUnsafe)
-	if latestUnsafe.Number < l2UnsafeTarget.Number {
-		logger.Warn("Latest unsafe block is older than target, using latest unsafe")
-		return latestUnsafe, nil
+	target := l2UnsafeTarget.Number
+
+	// Copy-paste from sort.Search, but also stops on and returns error
+	search := func(n int, f func(int) (bool, error)) (int, error) {
+		// Define f(-1) == false and f(n) == true.
+		// Invariant: f(i-1) == false, f(j) == true.
+		i, j := 0, n
+		for i < j {
+			h := int(uint(i+j) >> 1) // avoid overflow when computing h
+			// i ≤ h < j
+			ok, err := f(h)
+			if err != nil {
+				return -1, err
+			}
+			if !ok {
+				i = h + 1 // preserves f(i-1) == false
+			} else {
+				j = h // preserves f(j) == true
+			}
+		}
+		// i == j, f(i-1) == false, and f(j) (= f(i)) == true  =>  answer is i.
+		return i, nil
 	}
 
-	unsafeTarget, err := m.l2.L2BlockRefByHash(ctx, l2UnsafeTarget.Hash)
-	if err != nil {
-		return eth.L2BlockRef{}, err
+	// isInvalidBlock returns:
+	// - false: if the L1Origin field for the block belongs to the canonical chain
+	// - true: if the L1Origin field for the block is invalid/outdated/reorged
+	isInvalidBlock := func(blockNum uint64) (bool, error) {
+		current, err := m.l2.L2BlockRefByNumber(ctx, blockNum)
+		if err != nil {
+			logger.Error("Failed to get L2 block ref", "err", err, "blocknum", blockNum)
+			return true, err
+		}
+
+		// Check if L1Origin has been reorged
+		l1Blk, err := m.l1.L1BlockRefByNumber(ctx, current.L1Origin.Number)
+		if err != nil {
+			logger.Error("Failed to get L1 block ref", "err", err, "blocknum", current.L1Origin.Number)
+			return true, err
+		}
+		if l1Blk.Hash != current.L1Origin.Hash {
+			logger.Trace("L1Origin field is invalid/outdated, so block is invalid and should be reorged", "current.number", current.Number, "current.L1Origin", current.L1Origin, "new-L1Origin", l1Blk)
```

Comment:
I think this is better to be a debug log, it's probably not that common and rather unexpected.

Reply:
please also use more conventional logging field names, like
```suggestion
			logger.Debug("L1Origin field is invalid/outdated, so block is invalid and should be reorged", "currentNumber", current.Number, "currentL1Origin", current.L1Origin, "newL1Origin", l1Blk)
```